pkg/querier/frontend/worker.go

```diff
@@ -118,116 +114,80 @@ func (w *worker) watchDNSLoop(servCtx context.Context) error {
-				if cancel, ok := cancels[update.Addr]; ok {
-					cancel()
+				if mgr, ok := w.managers[update.Addr]; ok {
+					mgr.stop()
 				}
 
 			default:
 				return fmt.Errorf("unknown op: %v", update.Op)
 			}
 		}
+
+		w.resetParallelism()
 	}
 }
 
-// runMany starts N runOne loops for a given address.
-func (w *worker) runMany(ctx context.Context, address string) {
-	client, err := w.connect(address)
-	if err != nil {
-		level.Error(w.log).Log("msg", "error connecting", "addr", address, "err", err)
-		return
+func (w *worker) connect(ctx context.Context, address string) (FrontendClient, error) {
+	opts := []grpc.DialOption{
+		grpc.WithInsecure(),
+		grpc.WithBlock(),
 	}
+	opts = append(opts, w.cfg.GRPCClientConfig.DialOption([]grpc.UnaryClientInterceptor{middleware.ClientUserHeaderInterceptor}, nil)...)
 
-	w.wg.Add(w.cfg.Parallelism)
-	for i := 0; i < w.cfg.Parallelism; i++ {
-		go w.runOne(ctx, client)
+	conn, err := grpc.DialContext(ctx, address, opts...)
+	if err != nil {
+		return nil, err
 	}
+	return NewFrontendClient(conn), nil
 }
 
-// runOne loops, trying to establish a stream to the frontend to begin
-// request processing.
-func (w *worker) runOne(ctx context.Context, client FrontendClient) {
-	defer w.wg.Done()
-
-	backoff := util.NewBackoff(ctx, backoffConfig)
-	for backoff.Ongoing() {
-		c, err := client.Process(ctx)
-		if err != nil {
-			level.Error(w.log).Log("msg", "error contacting frontend", "err", err)
-			backoff.Wait()
-			continue
+func (w *worker) resetParallelism() {
+	addresses := make([]string, 0, len(w.managers))
+	for addr := range w.managers {
+		addresses = append(addresses, addr)
+	}
+	rand.Shuffle(len(addresses), func(i, j int) { addresses[i], addresses[j] = addresses[j], addresses[i] })
+
+	totalConcurrency := 0
+	for i, addr := range addresses {
+		concurrentRequests := 0
+		if w.cfg.MatchMaxConcurrency {
+			concurrentRequests = w.querierCfg.MaxConcurrent / len(w.managers)
+
+			// if max concurrency does not evenly divide into our frontends we will choose some
+			//  to receive an extra connection.  addresses were randomized above so this will be a
+			//  random selection of frontends
+			if i < w.querierCfg.MaxConcurrent%len(w.managers) {
+				level.Warn(w.log).Log("msg", "max concurrency is not evenly dividable across query frontends. adding an extra connection", "addr", addr)
+				concurrentRequests++
+			}
+		} else {
+			concurrentRequests = w.cfg.Parallelism
 		}
 
-		if err := w.process(c); err != nil {
-			level.Error(w.log).Log("msg", "error processing requests", "err", err)
-			backoff.Wait()
-			continue
+		// max concurrency is less than the total number of query frontends.  to prevent accidentally
+		// starving a frontend we are just going to always connect once to every frontend
+		//  this is dangerous b/c we may start exceeding promql max concurrency
+		if concurrentRequests == 0 {
+			concurrentRequests = 1
 		}
```

Comment:
Nit: I feel like the logic for the parallelisation per worker would make a nice self-contained function.